### PR TITLE
test: Migrate publish snapshotting to snapbox

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

This is part of #14039 and allows snapshotting in more places.

### Additional information

`InMemoryDir` is taken from some experiments I've been doing in snapshot for filesystem assertions.  That got held up because of design complexity but publish validation's needs are simpler than I was designing for.

### CODE REVIEW

File: crates/cargo-test-support/src/publish.rs [hunk 1]
```diff
@@ -216,15 +231,7 @@ pub fn validate_crate_contents(
...
         );
     }
-    if !expected_contents.is_empty() {
-        for (e_file_name, e_file_contents) in expected_contents {
-            let e_file_name = Path::new(e_file_name);
-            let actual_contents = files
-                .get(e_file_name)
-                .unwrap_or_else(|| panic!("file `{}` missing in archive", e_file_name.display()));
-            assert_match_exact(e_file_contents, actual_contents);
-        }
```
Comment (epage) [inline crates/cargo-test-support/src/publish.rs:215]: Something still missing is the reporting of which path failed

**Review by Theghost256 — APPROVED** [review]

File: crates/cargo-test-support/src/compare.rs [hunk 2]
```diff
@@ -706,6 +669,136 @@ impl fmt::Debug for WildStr<'_> {
...
+}
+
+impl From<()> for InMemoryDir {
+    fn from(_files: ()) -> Self {
+        let files = Vec::new();
+        Self { files }
+    }
+}
+
+macro_rules! impl_from_tuple_for_dirsnapshot {
```
Comment (weihanglo) [inline crates/cargo-test-support/src/compare.rs:767]: I know it is working, and I understand it is for supporting up to 9 tuples (right?). However the macro is a bit hard to read for me. Is this a de-facto decl macro pattern? Is it possible to document what it is doing?

**Review by weihanglo — APPROVED** [review]

**Review by weihanglo — APPROVED** [review]

File: crates/cargo-test-support/src/compare.rs (hunk 2 above)
Comment (weihanglo) [reply crates/cargo-test-support/src/compare.rs]: Thanks!

`r=me` when the version bump is fixed

**Pull request merged**